File: dotnet/src/Util/Order.cs
using MongoDB.Bson;
using MongoDB.Driver;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace MongoEncryption.Util
{
    internal static class Order
    {
        //Instance MongoCrypt class
        private static MongoCrypt _mongoCrypt = new MongoCrypt(Environment.GetEnvironmentVariable("mongoOrderConnectionString"), Environment.GetEnvironmentVariable("orderKeyName"));

        private static IMongoCollection<BsonDocument> OrderCollection(MongoClient client)
        {
            //Returns collection
            return client.GetDatabase(Environment.GetEnvironmentVariable("databaseName")).GetCollection<BsonDocument>(Environment.GetEnvironmentVariable("collectionName"));
        }

        public static async Task InsertOrderAsync(string input, bool encryptPII = true)
        {
            //Parse json provided and adds createData attribute
            var order = BsonDocument.Parse(input);
            order.Add("createDate", DateTime.UtcNow);

            if (encryptPII)
            {
                //Calls Explicity Client Encryption for PII data
                order["customerName"] = await _mongoCrypt.EncryptFieldAsync(order["customerName"], true);
                order["shippingAddress"] = await _mongoCrypt.EncryptFieldAsync(order["shippingAddress"]);
            }

            //Stores new order
            await OrderCollection(_mongoCrypt.AutoDecryptionClient).InsertOneAsync(order.ToBsonDocument());
        }

        public static async Task<List<dynamic>> FindOrderAsync(string field, BsonValue filterValue, bool encryptFilter = false, bool autodecrypt = true)
        {
            //Build query filter and encrypt search value if encryptFilter = true
            var filter = Builders<BsonDocument>.Filter.Eq(field, (encryptFilter ? await _mongoCrypt.EncryptFieldAsync(filterValue, true) : filterValue));

            //If autodecrypt == true, uses client connection with encryption settings to allow Auto Decryption. 
            //Otherwise, uses default connections and returns encrypted data as binary data.
            if (autodecrypt)
                return (await OrderCollection(_mongoCrypt.AutoDecryptionClient).FindAsync<dynamic>(filter)).ToList<dynamic>();
            else
            {
                var list = (await OrderCollection(_mongoCrypt.DefaultClient).FindAsync<BsonDocument>(filter)).ToList<BsonDocument>();

                var result = list.ConvertAll(new Converter<BsonDocument, dynamic>(ConvertBinary));

                return result;
            }
        }

        private static dynamic ConvertBinary(BsonDocument input)
        {
            Dictionary<string, dynamic> dict = new Dictionary<string, dynamic>();

            //Cast Binary to string
            foreach (var item in input.Elements)
            {
                if (item.Value.BsonType == BsonType.Binary)
                    dict.Add(item.Name, Convert.ToHexString(item.Value.AsByteArray));
                else
                    dict.Add(item.Name, item.Value);
            }

            return dict;
        }
    }
}

File: java/src/main/java/com/function/util/Order.java
package com.function.util;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import org.bson.BsonDateTime;
import org.bson.BsonDocument;
import org.bson.BsonInt32;
import org.bson.BsonString;
import org.bson.BsonValue;
import org.bson.Document;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.Filters;

public class Order {

    //Instance MongoCrypt class
    final private static MongoCrypt _mongoCrypt = new MongoCrypt(System.getenv("mongoOrderConnectionString"), System.getenv("orderKeyName"));

    private static MongoCollection<Document> OrderCollection(MongoClient client)
    {
        //Returns collection
        return client.getDatabase(System.getenv("databaseName")).getCollection(System.getenv("collectionName"));
    }

    public static void InsertOrder(String input)
    {
        final Boolean encryptPII = true;

        //Parse json provided and adds createData attribute
        var order = BsonDocument.parse(input);
        order.append("createDate", new BsonDateTime(Instant.now().toEpochMilli()));

        if (encryptPII)
        {
            //Calls Explicity Client Encryption for PII data
            order.replace("customerName", _mongoCrypt.EncryptField(order.get("customerName"), true));
            order.replace("shippingAddress", _mongoCrypt.EncryptField(order.get("shippingAddress"), false));
        }

        //Stores new order
        OrderCollection(_mongoCrypt.AutoDecryptionClient()).insertOne(Document.parse(order.toJson()));
    }

    public static List<Object> FindOrder(String field, Object inputFilter, Boolean encryptFilter, Boolean autodecrypt)
    {
        BsonValue filterValue = null;
        
        //Cast filter based on filter type. This sample expects Integer or String filters.
        if (inputFilter instanceof Integer)
            filterValue = new BsonInt32((Integer)inputFilter);
        else if (inputFilter instanceof String)
            filterValue = new BsonString((String)inputFilter);

        //Build query filter and encrypt search value if encryptFilter = true
        var filter = Filters.eq(field, (encryptFilter ? _mongoCrypt.EncryptField(filterValue, true) : filterValue));

        //If autodecrypt == true, uses client connection with encryption settings to allow Auto Decryption. 
        //Otherwise, uses default connections and returns encrypted data as binary data.
        if (autodecrypt)
            return OrderCollection(_mongoCrypt.AutoDecryptionClient()).find(filter).into(new ArrayList<>()); //map(doc -> BsonDocument.parse(doc.toJson()))
        else
            return OrderCollection(_mongoCrypt.DefaultClient()).find(filter).into(new ArrayList<>());
    }
}